framework/src/media/InputHandler.cpp

```diff
@@ -0,0 +1,342 @@
+								.build();
+
+		if (!streamBuffer) {
+			meddbg("streamBuffer is nullptr!\n");
+			return false;
+		}
+
+		setStreamBuffer(streamBuffer);
+	}
+
+	auto ret = mInputDataSource->open();
+	if (ret) {
+		registerDecoder(mInputDataSource->getAudioType(), mInputDataSource->getChannels(), mInputDataSource->getSampleRate());
+		start();
+	}
+	return ret;
+}
+
+bool InputHandler::close()
+{
+	stop();
+	unregisterDecoder();
+	return mInputDataSource->close();
+}
+
+ssize_t InputHandler::read(unsigned char *buf, size_t size)
+{
+	size_t rlen = 0;
+
+	start(); // Auto start
+
+	if (mBufferReader) {
+		rlen = mBufferReader->read(buf, size);
+	}
+
+	return (ssize_t)rlen;
+}
+
+bool InputHandler::start()
+{
+	if (!mInputDataSource->isPrepare()) {
+		return false;
+	}
+
+	createWorker();
+	return true;
+}
+
+bool InputHandler::stop()
+{
+	destroyWorker();
+	return true;
+}
+
+void InputHandler::createWorker()
+{
+	medvdbg("InputHandler::createWorker()\n");
+	if (mStreamBuffer && !mIsWorkerAlive) {
+		mStreamBuffer->reset();
+		mState = BUFFER_STATE_EMPTY;
+		mTotalBytes = 0;
+		mIsWorkerAlive = true;
+
+		long stackSize = CONFIG_INPUT_DATASOURCE_STACKSIZE;
+		pthread_attr_t attr;
+		pthread_attr_init(&attr);
+		pthread_attr_setstacksize(&attr, stackSize);
+		int ret = pthread_create(&mWorker, &attr, static_cast<pthread_startroutine_t>(InputHandler::workerMain), this);
+		if (ret != OK) {
+			meddbg("Fail to create DataSourceWorker thread, return value : %d\n", ret);
+			mIsWorkerAlive = false;
+			return;
+		}
+		pthread_setname_np(mWorker, "InputHandlerWorker");
+	}
+}
+
+void InputHandler::destroyWorker()
+{
+	medvdbg("InputHandler::destroyWorker()\n");
+	if (mIsWorkerAlive) {
+		// Setup flag,
+		mIsWorkerAlive = false;
+
+		// Worker may be blocked in buffer writing.
+		mBufferWriter->setEndOfStream();
+
+		// Wake worker up,
+		wakenWorker();
+
+		// Join thread.
+		pthread_join(mWorker, NULL);
+	}
+}
+
+void *InputHandler::workerMain(void *arg)
+{
+	auto stream = static_cast<InputHandler *>(arg);
+	auto worker = stream->mInputDataSource;
+
+	while (stream->mIsWorkerAlive) {
+		// Waken up by a reading/stopping operation
+		stream->sleepWorker();
+
+		// Worker may be stoped
+		if (!stream->mIsWorkerAlive) {
+			break;
+		}
+
+		auto size = stream->mBufferWriter->sizeOfSpace();
+		if (size > 0) {
+			auto buf = new unsigned char[size];
+			if (worker->read(buf, size) <= 0) {
+				// Error occurred, or inputting finished
+				stream->mBufferWriter->setEndOfStream();
+				delete[] buf;
+				break;
+			}
+
+			stream->writeToStreamBuffer(buf, size);
+			delete[] buf;
+		}
+	}
+
+	return NULL;
+}
+
+void InputHandler::sleepWorker()
+{
+	bool bEOS = mBufferReader->isEndOfStream();
+	size_t spaces = mBufferWriter->sizeOfSpace();
+
+	std::unique_lock<std::mutex> lock(mMutex);
+	// In case of EOS or overrun, sleep worker.
+	if (mIsWorkerAlive && (bEOS || (spaces == 0))) {
+		mCondv.wait(lock);
+	}
+}
+
+void InputHandler::wakenWorker()
```

Comment:
Is "waken" correct? or "wake" is correct?

Reply:
wake and waken are both right.